File: src/backtracking/cpp/generating_words/test/test_generating_words.cpp
#include "generating_words.h"
#include "gtest/gtest.h"

TEST(TestGeneratingWords, 3x3Board) {
  std::vector<std::vector<char>> board = {
      {'A', 'B', 'C'}, {'D', 'E', 'F'}, {'G', 'H', 'I'}};

  std::unordered_set<std::string> word_dict = {"ABC", "ABCD", "CEG", "BEHI",
                                               "XXX"};
  std::set<std::string> expected = {"ABC", "CEG", "BEHI"};

  std::set<std::string> actual = generate_words(board, word_dict);

  EXPECT_EQ(expected, actual);
}

TEST(TestGeneratingWords, EmptyBoard) {
  std::vector<std::vector<char>> board = {};
  std::unordered_set<std::string> word_dict = {};
  std::set<std::string> expected = {};

  std::set<std::string> actual = generate_words(board, word_dict);

  EXPECT_EQ(expected, actual);
}

TEST(TestGeneratingWords, EmptyWordDict) {
  std::vector<std::vector<char>> board = {
      {'A', 'B', 'C'}, {'D', 'E', 'F'}, {'G', 'H', 'I'}};
  std::unordered_set<std::string> word_dict = {};
  std::set<std::string> expected = {};

  std::set<std::string> actual = generate_words(board, word_dict);

  EXPECT_EQ(expected, actual);
}

File: src/backtracking/python/generating_words/tests/test_generating_words.py
from src.generate_words import generate_words
import unittest


class TestGenerateWords(unittest.TestCase):
    def test_3x3_board(self):
        board = (("A", "B", "C"), ("D", "E", "F"), ("G", "H", "I"))

        word_dict = ("ABC", "ABCD", "CEG", "BEHI", "XXX")
        expected = ("ABC", "CEG", "BEHI")
        actual = generate_words(board, word_dict)

        self.assertEqual(sorted(expected), sorted(actual))

    def test_empty_board(self):
        board = list()
        word_dict = list()
        expected = tuple()
        actual = generate_words(board, word_dict)

        self.assertEqual(sorted(expected), sorted(actual))

    def test_empty_word_dict(self):
        board = (("A", "B", "C"), ("D", "E", "F"), ("G", "H", "I"))
        word_dict = list()
        expected = tuple()
        actual = generate_words(board, word_dict)

        self.assertEqual(sorted(expected), sorted(actual))


if __name__ == "__main__":
    unittest.main()
